source/exercises/mailroom-oo.rst

```diff
@@ -51,29 +52,64 @@ One of the hardest parts of OO design (particularly in Python) is to know how "l
 
 Your class for the collection of donors will also hold the code that generates reports about multiple donors.
 
-Remember that the user interaction code (anything with an ``input`` function) should be outside of these "logic" or "model" classes.
 
-In general, you will have a method for each of the functions in your non-OO version. Which class they go it will depend on whether the method only needs the information from one donor, or from the whole collection.
+Command Line Interface
+^^^^^^^^^^^^^^^^^^^^^^
+
+Module responsible for main program flow (CLI - Command Line Interface).
+
+Let's call this module ``cli_main.py`` to represent the entry point for the mailroom program. This module will be using above classes we defined: ``Donor`` and ``DonorCollection``; it will also handle interaction with the user via the ``input`` function calls to gather user input and to provide the output to the console.
+
+What should go into this module?
+
+* the main "switch dictionary" to map user selection to the program features; in general, you will have a method for each of the mailroom functions.
+* ``input`` function calls to gather user input.
+* ``print`` statements to print to console.
+
+
+.. note::  Technically console print statement don't belong in your data classes, however for some features like send letters instead of "sending" we are simply printing so it is ok for this feature to reside in the data class.
+
+
+Why is this separation so important?
+
+The idea here is that we should be able to fairly easy replace this CLI program with a different type of interface like a GUI (Graphical User Interface) as an example and not having to make any changes to our data classes.
+The only thing that would need to change with a potential GUI feature is implementing actual gui elements to use those same classes.
+
+Test Driven Development
+-----------------------
 
-Rules of thumb for where to put methods:
+At this point we have done a great job refactoring the more complex code out of data holding classes and we are left with simple classes that are more straightforward to unit test.
 
-1) Hopefully, once you made your code testable, all the user-interaction code (with ``input()``) is self contained in functions that don't have any logic (data manipulation) in them. If not, then this is a good time to refactor.
+The ``Donor`` and ``DonorCollection`` classes should now have close to 100 percent code coverage.
 
-2) If a function does something with a single donor -- it should be a method in the Donor class.
+Although it is ok to have a single test file to test all of your functionality, it is more common practice to have a test file per module.
 
-3) If a function works with multiple donors -- it should be in the class that handles a collection of donors.
+For example, you may create 3 test files:
+``test_donor.py``, ``test_donor_collection.py`` and ``test_cli_main.py`` - each testing corresponding module.
```

Comment:
we may not want them to tryt o test the cli -- there isn't much to test without mocking input() -- and that's too much to ask at this stage (though I have had students figure it out).

I got into that in course two -- advanced testing.

Reply:
Ok I can remove test_cli_main reference as it should be optional at this point since it would be heavy on  mocks and although some students do use mocks in my py210, some are very confused about them 🙂